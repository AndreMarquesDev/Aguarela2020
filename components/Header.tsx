/* eslint-disable jsx-a11y/anchor-is-valid */
import React, { FC } from 'react';
import Link from 'next/link';
import classNames from 'classnames';
import logo from '../public/images/logo.png';

interface IHeaderProps {
    currentRoute: string;
}

const Header: FC<IHeaderProps> = props => {
    const {
        currentRoute,
    } = props;

    const aboutLinkStyles = classNames(
        'link',
        currentRoute === '/about' && 'active',
    );
    const projectsLinkStyles = classNames(
        'link',
        currentRoute === '/projects' && 'active',
    );
    const servicesLinkStyles = classNames(
        'link',
        currentRoute === '/services' && 'active',
    );
    const contactLinkStyles = classNames(
        'link',
        currentRoute === '/contact' && 'active',
    );

    return (
        <nav>
            <Link href="/">
                <a>
                    <img alt="Logo" src={logo} />
                </a>
            </Link>
            <ul>
                <li>
                    <Link href="/about">
                        <a className={aboutLinkStyles}>Sobre</a>
                    </Link>
                </li>
                <li>
                    <Link href="/projects">
                        <a className={projectsLinkStyles}>Projetos</a>
                    </Link>
                </li>
                <li>
                    <Link href="/services">
                        <a className={servicesLinkStyles}>Servi√ßos</a>
                    </Link>
                </li>
                <li>
                    <Link href="/contact">
                        <a className={contactLinkStyles}>Contacto</a>
                    </Link>
                </li>
            </ul>

            <style jsx>
                {`
                    nav {
                        display: flex;
                        align-items: center;
                        flex-direction: column;
                    }

                    img {
                        width: 150rem;
                    }

                    ul {
                        display: flex;
                        justify-content: center;
                        list-style: none;
                    }

                    li a {
                        font-size: 24rem;
                        margin: 0 10rem;
                        text-transform: uppercase;
                    }
                `}
            </style>
        </nav>
    );
};

export default Header;
