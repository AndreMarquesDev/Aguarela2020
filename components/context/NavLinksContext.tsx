import { createContext, Context } from 'react';

export interface NavLinksContextProps {
    setNavHeight?: React.Dispatch<React.SetStateAction<number>>;
}

const NavLinksContext: Context<NavLinksContextProps> = createContext(null);

export default NavLinksContext;
