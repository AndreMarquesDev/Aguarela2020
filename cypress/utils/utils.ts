import { Locale } from '../../types/Locale';
import {
    menuIconButtonDataTestId,
    nukaCarouselNextButtonDataTestId,
} from '../../utils/dataTestIds';
import { TextsInterface, textsPt, textsEn } from '../../utils/texts';

export const getLocalizedTexts = (locale: Locale): TextsInterface => {
    if (locale === Locale.Pt) {
        return textsPt;
    }

    return textsEn;
};

export const openMenuMobile = (isMobile: boolean, isFirefox: boolean): void => {
    if (isMobile) {
        cy.getByDataTestId(menuIconButtonDataTestId).click({ force: isFirefox });
    }
};

export const clickNextArrowButtonIfMobile = (
    isMobile: boolean,
    container: string,
    containerTitle: string
): void => {
    const isFirefox = Cypress.isBrowser('firefox');

    if (isMobile) {
        cy.getByDataTestIdParent(container, nukaCarouselNextButtonDataTestId).click({
            force: isFirefox,
        });

        cy.getByText(container, containerTitle).scrollIntoView();
    }
};
