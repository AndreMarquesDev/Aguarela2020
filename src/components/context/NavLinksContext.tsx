import type { Context } from 'react';
import { createContext } from 'react';

export type NavLinksContextProps = {
    isMenuOpen: boolean;
    toggleMenu: React.Dispatch<React.SetStateAction<boolean>>;
    setNavHeight?: React.Dispatch<React.SetStateAction<number>>;
};

// @ts-ignore
export const NavLinksContext: Context<NavLinksContextProps> = createContext(null);
