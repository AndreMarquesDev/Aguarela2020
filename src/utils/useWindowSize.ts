import { useEffect, useState } from 'react';

type WindowSize = {
    width: number;
    height: number;
};

export const useWindowSize = (): WindowSize => {
    const [windowSize, setWindowSize] = useState<WindowSize>({
        width: 0,
        height: 0,
    });

    useEffect(() => {
        const handleResize = (): void => {
            // eslint-disable-next-line react-hooks-extra/no-direct-set-state-in-use-effect
            setWindowSize({
                width: window.innerWidth,
                height: window.innerHeight,
            });
        };

        window.addEventListener('resize', handleResize);

        handleResize();

        return (): void => window.removeEventListener('resize', handleResize);
    }, []);

    return windowSize;
};

export enum Breakpoint {
    Desktop = 1280,
    Phablet = 1023,
    Tablet = 767,
    Mobile = 511,
}
