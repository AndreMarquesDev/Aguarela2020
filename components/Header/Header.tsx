import React, { useContext, useRef, useEffect } from 'react';
import Link from 'next/link';
import Image from 'next/image';
import { NavLinksContext } from '../context/NavLinksContext';
import { NavLinks } from '../NavLinks/NavLinks';
import { Page } from '../../utils/pages';
import { useWindowSize, Breakpoint } from '../../utils/useWindowSize';
import { MenuIcon } from '../MenuIcon/MenuIcon';
import { Locale } from '../../types/Locale';
import { headerDataTestId, homepageLogoLinkDataTestId } from '../../utils/dataTestIds';

export interface HeaderProps {
    currentRoute: string;
    language: Locale;
}

export const Header = ({ currentRoute, language }: HeaderProps): JSX.Element => {
    const { setNavHeight } = useContext(NavLinksContext);

    const navRef = useRef(null);

    useEffect(() => {
        if (setNavHeight) {
            setNavHeight(navRef.current.clientHeight);
        }
    });

    const { isMenuOpen, toggleMenu } = useContext(NavLinksContext);
    const handleMenuClick = (): void => toggleMenu(!isMenuOpen);

    const windowSize = useWindowSize();
    const isDesktop = windowSize.width > Breakpoint.Tablet;
    const isPhablet = windowSize.width < Breakpoint.Phablet;

    const logoWidth = isPhablet ? 150 : 250;

    return (
        <header data-testid={headerDataTestId}>
            <nav ref={navRef} className="wrapper">
                <Link href={`/${language}/${'homepage' as Page}`} prefetch={false}>
                    {/* eslint-disable-next-line jsx-a11y/anchor-is-valid */}
                    <a data-testid={homepageLogoLinkDataTestId}>
                        <Image alt="Logo" height={35} src="/images/logo.svg" width={logoWidth} />
                    </a>
                </Link>

                <NavLinks
                    currentRoute={currentRoute}
                    isMenuOpen={isMenuOpen}
                    isMobile={!isDesktop}
                    language={language}
                />
                <MenuIcon isOpen={isMenuOpen} isVisible={!isDesktop} onClick={handleMenuClick} />
            </nav>

            <style jsx>
                {`
                    @import './styles/_vars.scss';

                    nav {
                        display: flex;
                        align-items: center;
                        justify-content: space-between;
                        padding-top: 60rem;
                        padding-bottom: 60rem;
                    }

                    @include tablet {
                        nav {
                            padding-top: 30rem;
                            padding-bottom: 30rem;
                        }
                    }
                `}
            </style>
        </header>
    );
};
