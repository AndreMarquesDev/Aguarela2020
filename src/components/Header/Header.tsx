import type { Locale } from '../../types/Locale';
import type { Page } from '../../utils/pages';
import Image from 'next/image';
import Link from 'next/link';
import React, { useContext, useEffect, useRef } from 'react';
import { tabletMediaQuery } from '../../styles/mediaQueries';
import { headerDataTestId, homepageLogoLinkDataTestId } from '../../utils/dataTestIds';
import { Breakpoint, useWindowSize } from '../../utils/useWindowSize';
import { NavLinksContext } from '../context/NavLinksContext';
import { MenuIcon } from '../MenuIcon/MenuIcon';
import { NavLinks } from '../NavLinks/NavLinks';

export type HeaderProps = {
    currentRoute: string;
    language: Locale;
};

export const Header = ({ currentRoute, language }: HeaderProps): JSX.Element => {
    const { setNavHeight } = useContext(NavLinksContext);

    const navRef = useRef<HTMLDivElement>(null);

    useEffect(() => {
        if (setNavHeight && navRef.current) {
            setNavHeight(navRef.current.clientHeight);
        }
    });

    const { isMenuOpen, toggleMenu } = useContext(NavLinksContext);
    const handleMenuClick = (): void => toggleMenu(!isMenuOpen);

    const windowSize = useWindowSize();
    const isDesktop = windowSize.width > Breakpoint.Tablet;
    const isPhablet = windowSize.width < Breakpoint.Phablet;

    const logoWidth = isPhablet ? 150 : 250;

    return (
        <header data-testid={headerDataTestId}>
            <nav ref={navRef} className="wrapper">
                <Link
                    data-testid={homepageLogoLinkDataTestId}
                    href={`/${language}/${'homepage' as Page}`}
                    prefetch={false}
                >
                    <Image alt="Logo" height={35} src="/images/logo.svg" width={logoWidth} />
                </Link>

                <NavLinks
                    currentRoute={currentRoute}
                    isMenuOpen={isMenuOpen}
                    isMobile={!isDesktop}
                    language={language}
                />
                <MenuIcon isOpen={isMenuOpen} isVisible={!isDesktop} onClick={handleMenuClick} />
            </nav>

            <style jsx>
                {`
                    nav {
                        display: flex;
                        align-items: center;
                        justify-content: space-between;
                        padding-top: 60rem;
                        padding-bottom: 60rem;

                        @media (${tabletMediaQuery}) {
                            padding-top: 30rem;
                            padding-bottom: 30rem;
                        }
                    }
                `}
            </style>
        </header>
    );
};
