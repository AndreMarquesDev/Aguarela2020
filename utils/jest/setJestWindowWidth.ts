import { Breakpoint } from '../useWindowSize';

export const setJestWindowWidth = (breakpoint: Breakpoint): void => {
    let width;

    if (breakpoint === Breakpoint.Mobile) {
        width = Breakpoint.Mobile - 1;
    } else if (breakpoint === Breakpoint.Phablet) {
        width = Breakpoint.Phablet - 1;
    } else {
        width = Breakpoint.Desktop + 1;
    }

    Object.defineProperty(window, 'innerWidth', {
        writable: true,
        configurable: true,
        value: width,
    });
};
