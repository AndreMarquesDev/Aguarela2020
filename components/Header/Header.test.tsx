import '@testing-library/jest-dom';
import React from 'react';
import { render, RenderResult, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import { Header, HeaderProps } from './Header';
import NavLinksContext from '../context/NavLinksContext';
import { Breakpoint } from '../../utils/useWindowSize';
import { setJestWindowWidth } from '../../utils/jest/setJestWindowWidth';

const baseProps: HeaderProps = {
    currentRoute: 'services',
    language: 'pt',
};

const mockSetNavHeight = jest.fn();
const mockToggleMenu = jest.fn();

const renderComponent = (
    setNavHeight: jest.Mock,
    newProps?: Partial<HeaderProps>
): RenderResult => {
    return render(
        <NavLinksContext.Provider
            value={{ isMenuOpen: false, toggleMenu: mockToggleMenu, setNavHeight }}
        >
            <Header {...baseProps} {...newProps} />
        </NavLinksContext.Provider>
    );
};

describe('<Header />', () => {
    test('renders properly', () => {
        const { container } = renderComponent(mockSetNavHeight);

        expect(mockSetNavHeight).toHaveBeenCalledTimes(2);

        expect(container).toMatchSnapshot();
    });

    test('renders properly below phablet', () => {
        setJestWindowWidth(Breakpoint.Mobile);

        const { container } = renderComponent(mockSetNavHeight);

        expect(container).toMatchSnapshot();
    });

    test('does not call the "setNavHeight" function if it is undefined', () => {
        renderComponent(undefined);

        expect(mockSetNavHeight).toHaveBeenCalledTimes(0);
    });

    test('calls the "toggleMenu" function when clicking on <MenuIcon />', () => {
        renderComponent(mockSetNavHeight);

        const menuIcon = screen.getByRole('button');

        expect(menuIcon).toBeInTheDocument();

        expect(mockToggleMenu).toHaveBeenCalledTimes(0);

        userEvent.click(menuIcon);

        expect(mockToggleMenu).toHaveBeenCalledTimes(1);
    });
});
