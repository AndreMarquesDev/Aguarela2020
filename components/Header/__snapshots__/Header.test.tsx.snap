// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Header /> renders properly 1`] = `
<div>
  <header
    class="jsx-2230556198"
  >
    <nav
      class="jsx-2230556198 wrapper"
    >
      <a
        class="jsx-2230556198"
        href="/pt/homepage"
      >
        <div
          style="display: inline-block; max-width: 100%; overflow: hidden; position: relative; box-sizing: border-box; margin: 0px;"
        >
          <div
            style="box-sizing: border-box; display: block; max-width: 100%;"
          >
            <img
              alt=""
              aria-hidden="true"
              role="presentation"
              src="data:image/svg+xml;base64,PHN2ZyB3aWR0aD0iMjUwIiBoZWlnaHQ9IjM1IiB4bWxucz0iaHR0cDovL3d3dy53My5vcmcvMjAwMC9zdmciIHZlcnNpb249IjEuMSIvPg=="
              style="max-width: 100%; display: block; margin: 0px; padding: 0px;"
            />
          </div>
          <img
            alt="Logo"
            decoding="async"
            src="data:image/gif;base64,R0lGODlhAQABAIAAAAAAAP///yH5BAEAAAAALAAAAAABAAEAAAIBRAA7"
            style="visibility: hidden; position: absolute; top: 0px; left: 0px; bottom: 0px; right: 0px; box-sizing: border-box; padding: 0px; margin: auto; display: block; width: 0px; height: 0px; min-width: 100%; max-width: 100%; min-height: 100%; max-height: 100%;"
          />
        </div>
      </a>
      <ul
        class="jsx-1064030420 "
      >
        <li
          class="jsx-1064030420"
        >
          <a
            class="jsx-1064030420 "
            href="/pt/about"
          >
            Sobre
          </a>
        </li>
        <li
          class="jsx-1064030420"
        >
          <a
            class="jsx-1064030420 "
            href="/pt/projects"
          >
            Projetos
          </a>
        </li>
        <li
          class="jsx-1064030420"
        >
          <a
            class="jsx-1064030420 active"
            href="/pt/services"
          >
            Serviços
          </a>
        </li>
        <li
          class="jsx-1064030420"
        >
          <a
            class="jsx-1064030420 "
            href="/pt/contact"
          >
            Contacto
          </a>
        </li>
        <li
          class="jsx-1064030420 languageButton"
        >
          <button
            class="jsx-1698987686"
            type="button"
          >
            pt
          </button>
        </li>
      </ul>
      <i
        aria-label="toggle menu"
        class="jsx-2133906697 menuIcon hidden"
        role="button"
        tabindex="0"
      >
        <span
          class="jsx-2133906697"
        />
      </i>
    </nav>
  </header>
</div>
`;

exports[`<Header /> renders properly below phablet 1`] = `
<div>
  <header
    class="jsx-2230556198"
  >
    <nav
      class="jsx-2230556198 wrapper"
    >
      <a
        class="jsx-2230556198"
        href="/pt/homepage"
      >
        <div
          style="display: inline-block; max-width: 100%; overflow: hidden; position: relative; box-sizing: border-box; margin: 0px;"
        >
          <div
            style="box-sizing: border-box; display: block; max-width: 100%;"
          >
            <img
              alt=""
              aria-hidden="true"
              role="presentation"
              src="data:image/svg+xml;base64,PHN2ZyB3aWR0aD0iMTUwIiBoZWlnaHQ9IjM1IiB4bWxucz0iaHR0cDovL3d3dy53My5vcmcvMjAwMC9zdmciIHZlcnNpb249IjEuMSIvPg=="
              style="max-width: 100%; display: block; margin: 0px; padding: 0px;"
            />
          </div>
          <img
            alt="Logo"
            decoding="async"
            src="data:image/gif;base64,R0lGODlhAQABAIAAAAAAAP///yH5BAEAAAAALAAAAAABAAEAAAIBRAA7"
            style="visibility: hidden; position: absolute; top: 0px; left: 0px; bottom: 0px; right: 0px; box-sizing: border-box; padding: 0px; margin: auto; display: block; width: 0px; height: 0px; min-width: 100%; max-width: 100%; min-height: 100%; max-height: 100%;"
          />
        </div>
      </a>
      <ul
        class="jsx-1064030420 mobileLayout"
      >
        <li
          class="jsx-1064030420"
        >
          <a
            class="jsx-1064030420 "
            href="/pt/about"
          >
            Sobre
          </a>
        </li>
        <li
          class="jsx-1064030420"
        >
          <a
            class="jsx-1064030420 "
            href="/pt/projects"
          >
            Projetos
          </a>
        </li>
        <li
          class="jsx-1064030420"
        >
          <a
            class="jsx-1064030420 active"
            href="/pt/services"
          >
            Serviços
          </a>
        </li>
        <li
          class="jsx-1064030420"
        >
          <a
            class="jsx-1064030420 "
            href="/pt/contact"
          >
            Contacto
          </a>
        </li>
        <li
          class="jsx-1064030420 languageButton white"
        >
          <button
            class="jsx-1698987686"
            type="button"
          >
            pt
          </button>
        </li>
      </ul>
      <i
        aria-label="toggle menu"
        class="jsx-2133906697 menuIcon"
        role="button"
        tabindex="0"
      >
        <span
          class="jsx-2133906697"
        />
      </i>
    </nav>
  </header>
</div>
`;
